blocks/blocks.go

```diff
@@ -41,46 +39,70 @@ type blockValidator interface {
 
 // BlockHandler is the struct responsible for storing meta data needed to process blocks from gossip.
 type BlockHandler struct {
-	log.Log
-	fetcher     system.Fetcher
-	traverse    forBlockInView
-	depth       uint32
-	mesh        mesh
-	validator   blockValidator
-	goldenATXID types.ATXID
+	logger log.Log
+	cfg    Config
+
+	fetcher   system.Fetcher
+	mesh      mesh
+	validator blockValidator
 }
 
 // Config defines configuration for block handler.
 type Config struct {
-	Depth       uint32
-	GoldenATXID types.ATXID
+	MaxExceptions int
+}
+
+// defaultConfig fro BlockHandler.
+func defaultConfig() Config {
+	return Config{
+		MaxExceptions: 1000,
+	}
+}
+
+// Opt for configuring BlockHandler.
+type Opt func(b *BlockHandler)
+
+// WithMaxExceptions defines max allowed exceptions in a block.
+func WithMaxExceptions(max int) Opt {
+	return func(b *BlockHandler) {
+		b.cfg.MaxExceptions = max
+	}
+}
+
+// WithLogger defines logger for BlockHandler.
+func WithLogger(logger log.Log) Opt {
+	return func(b *BlockHandler) {
+		b.logger = logger
+	}
 }
 
 // NewBlockHandler creates new BlockHandler.
-func NewBlockHandler(cfg Config, fetcher system.Fetcher, m mesh, v blockValidator, lg log.Log) *BlockHandler {
-	return &BlockHandler{
-		Log:         lg,
-		fetcher:     fetcher,
-		traverse:    m.ForBlockInView,
-		depth:       cfg.Depth,
-		mesh:        m,
-		validator:   v,
-		goldenATXID: cfg.GoldenATXID,
+func NewBlockHandler(fetcher system.Fetcher, m mesh, v blockValidator, opts ...Opt) *BlockHandler {
+	b := &BlockHandler{
+		logger:    log.NewNop(),
+		cfg:       defaultConfig(),
+		fetcher:   fetcher,
+		mesh:      m,
+		validator: v,
+	}
+	for _, opt := range opts {
+		opt(b)
```

Comment:
i always wonder why some args are configured and some now when most of them will be manipulated during testing.

what's your rationale not make all arguments to BlockHandler as options?

Reply:
i think there should be no "required" options, so i am passing all dependencies explicitly. if one of those deps is not provided block handler will crash with nil-pointer error.